### PR TITLE
Refactor GcnEvent Handler

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR refactors the GcnEvent handler so we can use it in the gcn_service.

### CODE REVIEW

File: skyportal/handlers/api/gcn.py [hunk 1]
```diff
@@ -33,20 +33,25 @@
 Session = scoped_session(sessionmaker(bind=DBSession.session_factory.kw["bind"]))
 
 
-def post_gcnevent(payload, user_id):
+def post_gcnevent(payload, user_id, session=Session()):
```
Comment (stefanv) [inline skyportal/handlers/api/gcn.py:36]:
```suggestion
def post_gcnevent(payload, user_id, session):
```

I don't think a session should be instantiated like that.

Comment (mcoughlin) [reply skyportal/handlers/api/gcn.py]: Yes good point. Fixed.

File: services/gcn_service/gcn_service.py [hunk 2]
```diff
@@ -1,23 +1,20 @@
...
-        return token
-    except (FileNotFoundError, TypeError, KeyError):
-        print('Error: no token specified, and no suitable token found in .tokens.yaml')
-        return None
+env, cfg = load_env()
 
+init_db(**cfg['database'])
 
-admin_token = get_token()
+Session = scoped_session(sessionmaker(bind=DBSession.session_factory.kw["bind"]))
```
Comment (stefanv) [inline services/gcn_service/gcn_service.py:17]: Could probably use DBSession directly, curious why this is necessary.

File: services/gcn_service/gcn_service.py [hunk 3]
```diff
@@ -36,9 +33,10 @@ def get_token():
     gcn.NoticeType.ICECUBE_ASTROTRACK_BRONZE,
 )
 def handle(payload, root):
-    response_status, data = api(
-        'POST', 'gcn_event', data={'xml': payload}, token=admin_token
-    )
+
+    session = Session()
```
Comment (stefanv) [inline services/gcn_service/gcn_service.py:37]: Use `with Session()` syntax?

File: skyportal/handlers/api/gcn.py [hunk 4]
```diff
@@ -69,82 +162,10 @@ def post(self):
             return self.error("xml must be present in data to parse GcnEvent")
 
         payload = data['xml']
+        session = Session()
```
Comment (stefanv) [inline skyportal/handlers/api/gcn.py:165]: with Session

File: services/gcn_service/gcn_service.py (hunk 2 above)
Comment (mcoughlin) [reply services/gcn_service/gcn_service.py]: this was just copying from gcn.py, I removed.

File: services/gcn_service/gcn_service.py (hunk 3 above)
Comment (mcoughlin) [reply services/gcn_service/gcn_service.py]: added.

File: skyportal/handlers/api/gcn.py (hunk 4 above)
Comment (mcoughlin) [reply skyportal/handlers/api/gcn.py]: added.

**Review by stefanv — APPROVED** [review]

File: services/gcn_service/gcn_service.py [hunk 5]
```diff
@@ -1,23 +1,17 @@
...
-        token = yaml.load(open('.tokens.yaml'), Loader=yaml.Loader)['INITIAL_ADMIN']
-        print('Token loaded from `.tokens.yaml`')
-        return token
-    except (FileNotFoundError, TypeError, KeyError):
-        print('Error: no token specified, and no suitable token found in .tokens.yaml')
-        return None
+from skyportal.handlers.api.gcn import post_gcnevent
+from skyportal.models import (
+    DBSession,
+)
```
Comment (guynir42) [inline services/gcn_service/gcn_service.py:10]:
```suggestion
from skyportal.models import DBSession
```

File: skyportal/handlers/api/gcn.py [hunk 6]
```diff
@@ -32,6 +33,98 @@
...
+            .filter_by(
+                dateobs=dateobs,
+                localization_name=skymap["localization_name"],
+            )
+            .one()
+        )
+    except NoResultFound:
+        localization = Localization(**skymap)
+        session.add(localization)
+        session.commit()
```
Comment (guynir42) [inline skyportal/handlers/api/gcn.py:118]: The only place where you are committing is inside an except clause. That doesn't look right...

Comment (mcoughlin) [reply skyportal/handlers/api/gcn.py:118]: This is just checking for a repeat of the localization.

Comment (guynir42) [reply skyportal/handlers/api/gcn.py:118]: Yeah but you are adding and modifying some rows in this function. Are you sure those changes are getting committed at all?

Comment (mcoughlin) [reply skyportal/handlers/api/gcn.py:118]: @guynir42 its behaved that way for a long time now. I tested it repeats the old behavior.

Comment (stefanv) [reply skyportal/handlers/api/gcn.py:118]: Hrm, what needs to be clarified is if each of the following sometimes raises a NoResultsFound exception:

```python
         event = GcnEvent.query.filter_by(dateobs=dateobs).one()
```

and

```python
            Localization.query_records_accessible_by(
                user,
            )
            .filter_by(
                dateobs=dateobs,
                localization_name=skymap["localization_name"],
            )
            .one()
```

Comment (mcoughlin) [reply skyportal/handlers/api/gcn.py:118]: Given that the demo data and the usual flow works, my intuition is this is all fine.

**Review by guynir42 — APPROVED** [review]

**Pull request merged**